src/scp/QuorumIntersectionChecker.cpp

```diff
@@ -0,0 +1,1145 @@
+    BitSet mMaxSCC;
+
+    QBitSet convertSCPQuorumSet(SCPQuorumSet const& sqs);
+    void buildGraph(QuorumTracker::QuorumMap const& qmap);
+    void buildSCCs();
+
+    bool containsQuorumSlice(BitSet const& bs, QBitSet const& qbs) const;
+    bool containsQuorumSliceForNode(BitSet const& bs, size_t node) const;
+    BitSet contractToMaximalQuorum(BitSet nodes) const;
+    bool isAQuorum(BitSet const& nodes) const;
+    bool isMinimalQuorum(BitSet const& nodes) const;
+    bool hasDisjointQuorum(BitSet const& nodes) const;
+    void noteFoundDisjointQuorums(BitSet const& nodes,
+                                  BitSet const& disj) const;
+    std::string nodeName(size_t node) const;
+
+    friend class MinQuorumEnumerator;
+
+public:
+    QuorumIntersectionCheckerImpl(QuorumTracker::QuorumMap const& qmap,
+                                  Config const& cfg)
+        : mCfg(cfg),
+          mLogTrace(Logging::logTrace("SCP")),
+          mTSC(mGraph)
+    {
+        buildGraph(qmap);
+        buildSCCs();
+    }
+
+    bool networkEnjoysQuorumIntersection() const override;
+
+    std::pair<std::vector<PublicKey>,
+              std::vector<PublicKey>>
+    getPotentialSplit() const override { return mPotentialSplit; }
+
+};
+
+std::shared_ptr<QuorumIntersectionChecker>
+QuorumIntersectionChecker::create(QuorumTracker::QuorumMap const& qmap,
+                                  Config const& cfg)
+{
+    return std::make_shared<QuorumIntersectionCheckerImpl>(qmap, cfg);
+}
+
+void
+QuorumIntersectionCheckerImpl::Stats::log() const
+{
+    CLOG(DEBUG, "SCP")
+        << "Quorum intersection checker stats:";
+    size_t exits = (mEarlyExit1s +
+                    mEarlyExit21s +
+                    mEarlyExit22s +
+                    mEarlyExit31s +
+                    mEarlyExit32s);
+    CLOG(DEBUG, "SCP")
+        << "[Nodes: " << mTotalNodes
+        << ", SCCs: " << mNumSCCs
+        << ", MaxSCC: " << mMaxSCC
+        << ", MaxQs:" << mMaxQuorumsSeen
+        << ", MinQs:" << mMinQuorumsSeen
+        << ", Calls:" << mCallsStarted
+        << ", Terms:" << mTerminations
+        << ", Exits:" << exits
+        << "]";
+    CLOG(DEBUG, "SCP")
+        << "Detailed exit stats:";
+    CLOG(DEBUG, "SCP")
+        << "[X1:" << mEarlyExit1s
+        << ", X2.1:" << mEarlyExit21s
+        << ", X2.2:" << mEarlyExit22s
+        << ", X3.1:" << mEarlyExit31s
+        << ", X3.2:" << mEarlyExit32s
+        << "]";
+}
+
+// This function is the innermost call in the checker and must be as fast
+// as possible. We spend almost all of our time in here.
+bool
+QuorumIntersectionCheckerImpl::containsQuorumSlice(BitSet const& bs,
+                                               QBitSet const& qbs) const
+{
+    // First we do a very quick check: do we have enough bits in 'bs'
+    // intersected with the top-level set of nodes to meet the threshold for
+    // this qset?
+    size_t intersecting = bs.intersection_count(qbs.mNodes);
```

Comment:
Can you use the intersection `ibs = bs & qbs.mAllSuccessors` instead of `bs` as it may reduce the size of loops all around (assuming we often don't have much intersection in most cases)?

Reply:
Read a little further down: we do use the mAllSuccessors set for the 3rd (negative) test before the loop.